pubsublite-kafka-shim/src/main/java/com/google/cloud/pubsublite/kafka/SingleSubscriptionConsumerImpl.java

```diff
@@ -0,0 +1,288 @@
+  @GuardedBy("monitor.monitor")
+  private boolean wakeupTriggered = false;
+
+  SingleSubscriptionConsumerImpl(
+      TopicPath topic,
+      boolean autocommit,
+      PullSubscriberFactory subscriberFactory,
+      CommitterFactory committerFactory) {
+    this.topic = topic;
+    this.autocommit = autocommit;
+    this.subscriberFactory = subscriberFactory;
+    this.committerFactory = committerFactory;
+  }
+
+  @Override
+  public void setAssignment(Set<Partition> assignment) {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      List<SubscriberState> unassigned =
+          ImmutableSet.copyOf(partitions.keySet()).stream()
+              .filter(p -> !assignment.contains(p))
+              .map(p -> partitions.remove(p))
+              .collect(Collectors.toList());
+      for (SubscriberState state : unassigned) {
+        state.subscriber.close();
+        state.committer.stopAsync().awaitTerminated();
+      }
+      assignment.stream()
+          .filter(p -> !partitions.containsKey(p))
+          .forEach(
+              ExtractStatus.rethrowAsRuntime(
+                  partition -> {
+                    SubscriberState s = new SubscriberState();
+                    s.subscriber =
+                        subscriberFactory.newPullSubscriber(
+                            partition,
+                            SeekRequest.newBuilder()
+                                .setNamedTarget(NamedTarget.COMMITTED_CURSOR)
+                                .build());
+                    s.committer = committerFactory.newCommitter(partition);
+                    s.committer.startAsync().awaitRunning();
+                    partitions.put(partition, s);
+                  }));
+    } catch (Throwable t) {
+      throw ExtractStatus.toCanonical(t).getStatus().asRuntimeException();
+    }
+  }
+
+  @Override
+  public Set<Partition> assignment() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return partitions.keySet();
+    }
+  }
+
+  @Override
+  public ConsumerRecords<byte[], byte[]> poll(Duration duration) {
+    Map<Partition, Queue<SequencedMessage>> partitionQueues = new HashMap<>();
+    try {
+      if (wakeupTriggered) throw new WakeupException();
+      while (!duration.isZero()) {
+        Duration sleepFor = Collections.min(ImmutableList.of(duration, Duration.ofMillis(100)));
+        if (wakeupTriggered) throw new WakeupException();
+        Thread.sleep(sleepFor.toMillis());
+        duration = duration.minus(sleepFor);
+        try (CloseableMonitor.Hold h = monitor.enter()) {
+          partitions.forEach(
+              ExtractStatus.rethrowAsRuntime(
+                  (partition, state) -> {
+                    partitionQueues
+                        .computeIfAbsent(partition, x -> new ArrayDeque<>())
+                        .addAll(state.subscriber.pull());
+                  }));
+        }
+      }
+      try (CloseableMonitor.Hold h = monitor.enter()) {
```

Comment:
Do we need this last read?

Reply:
Yes. Added an explanatory comment as to why.